### PR TITLE
Refactor `io` crate

### PR DESCRIPTION
**Pull request by tcharding**
Do various refactorings to the `io` crate. These are all trivial to review except the bigish commit: `d3d3e50f Move error code to submodule` which is code move only.

### CODE REVIEW

**Comment by tcharding** [comment]
Oh, I might be out of line here. Code in `consensus::encode` uses the same whitespace style as the original `io` code.

File: io/src/lib.rs [hunk 1]
```diff
@@ -13,9 +13,9 @@
 //! This traits are not one-for-one drop-ins, but are as close as possible while still implementing
 //! `std::io`'s traits without unnecessary complexity.
 
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (Kixunil) [inline io/src/lib.rs:16]: FTR I'm used to unconditionally setting `no_std` and conditionally `extern crate std`. I'm not sure what all pros and cons are but at least I find `#![no_std]` more readable than this thing.

**Review by Kixunil — APPROVED** [review]
ACK 97b0030e215f8e807127308650c1ce00513e3cc8

**Comment by apoelstra** [comment]
Needs rebase.

**Review by Kixunil — APPROVED** [review]
ACK b1870656c945d2417c55d46e8ef371b30776f0ba

**Review by apoelstra — APPROVED** [review]
ACK b1870656c945d2417c55d46e8ef371b30776f0ba

**Pull request merged**